packages/gestalt/src/SideNavigation/navigationChildrenUtils.js

```diff
@@ -0,0 +1,113 @@
+// @flow strict
+import { type Node as ReactNode } from 'react';
+import classnames from 'classnames';
+import ItemsEllipsis, { type Props as EllipsisProps } from './ItemsEllipsis';
+import styles from '../SideNavigation.css';
+import flattenChildren, { type ReactChildArray } from '../utils/flattenChildren';
+
+export const ALLOWED_CHILDREN_MAP = {
+  main: ['SideNavigation.Section', 'SideNavigation.TopItem', 'SideNavigation.Group'],
+  nested: ['SideNavigation.NestedItem', 'SideNavigation.NestedGroup'],
+};
+
+export function validateChildren({
+  children,
+  filterLevel,
+}: {
+  children: ReactChildArray,
+  filterLevel: 'main' | 'nested',
+}) {
+  children.forEach((child) => {
+    const isTopLevel = filterLevel === 'main';
+
+    if (!ALLOWED_CHILDREN_MAP[filterLevel].includes(child.type.displayName)) {
+      throw new Error(
+        `Gestalt ${child.type.displayName} cannot be used at ${
+          isTopLevel ? 'the top' : 'a nested'
+        } level`,
+      );
+    }
+  });
+}
+
+export function countItemsWithIcon(children: ReactChildArray): number {
```

Comment:
why do we need this function?

Reply:
nit can you add a description to each helper function?

Reply:
sections or items collapse based on icons. sidenav collapses as empty when there is no icons. that's why we need to count icons.